historysync.go

```diff
@@ -195,7 +192,69 @@ func (portal *Portal) convertBackfill(messages []*imessage.Message) ([]*event.Ev
 		}
 		isRead = msg.IsRead || msg.IsFromMe || (unreadThreshold >= 0 && time.Since(msg.Time) > unreadThreshold)
 	}
-	return events, metas, metaIndexes, isRead, nil
+	return events, metas, metaIndexes, isRead, lastMessage, nil
+}
+
+func (portal *Portal) convertTapbacks(messages []*imessage.Message) ([]*event.Event, []messageWithIndex, map[messageIndex]int, bool, *imessage.Message, error) {
+	events := make([]*event.Event, 0, len(messages))
+	metas := make([]messageWithIndex, 0, len(messages))
+	metaIndexes := make(map[messageIndex]int, len(messages))
+	unreadThreshold := time.Duration(portal.bridge.Config.Bridge.Backfill.UnreadHoursThreshold) * time.Hour
+	var isRead bool
+	var lastMessage *imessage.Message
+	for _, msg := range messages {
+		//Only want tapbacks
+		if msg.Tapback == nil {
+			continue
+		}
+
+		//Skip the last message in the array, we will add it later to correct inbox sorting
+		if msg == messages[len(messages)-1] && len(messages) > 1 /* we call this function again with one element in the array for the last message, so we'll want to process it */ {
+			portal.log.Debugln("Skipping message", msg.GUID, "in backfill, last one in the convo")
+			lastMessage = msg
+			continue
+		}
+
+		intent := portal.getIntentForMessage(msg, nil)
+		dbMessage := portal.bridge.DB.Message.GetByGUID(portal.GUID, msg.Tapback.TargetGUID, msg.Tapback.TargetPart)
+		if dbMessage == nil {
+			//TODO BUG: This occurs when trying to find the target reaction for a rich link, related to #183
+			portal.log.Errorfln("Failed to get target message for tabpack, %+v", msg)
+			continue
+		}
+
+		evt := &event.Event{
+			Sender:    intent.UserID,
+			Type:      event.EventReaction,
+			Timestamp: msg.Time.UnixMilli(),
+			Content: event.Content{
+				Parsed: &event.ReactionEventContent{
+					RelatesTo: event.RelatesTo{
+						Type:    event.RelAnnotation,
+						EventID: dbMessage.MXID,
+						Key:     msg.Tapback.Type.Emoji(),
+					},
+				},
+			},
+		}
+
+		var err error
+		evt.Type, err = portal.encrypt(intent, &evt.Content, evt.Type)
+		if err != nil {
+			return nil, nil, nil, false, nil, err
+		}
+		intent.AddDoublePuppetValue(&evt.Content)
```

Comment:
This shouldn't be here at all (and might have been causing the inbox issues if the sorting was actually wrong)